accepted/single-file/design.md

```diff
@@ -66,225 +57,211 @@ Publishing to a single file can be triggered by adding the following property to
+* Throw an `AssemblyLoadedFromBundle` exception
+* The empty string, similar to assemblies loaded from [byte-array](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.location?view=netcore-3.1#property-value),
 * The simple name of the assembly (with no path).
-* A special UNC notation such as `<bundle-path>/:/asm.dll` to denote files that come from the bundle. 
 * The path of the assembly as if it were not to be packaged into the single-file.
-* In the case of files spilled to disk (ex: ready-to-run assemblies) -- the actual location of the extracted file.
+* A special UNC notation such as `<bundle-path>/:/asm.dll` to denote files that come from the bundle. 
 * A configurable selection of the above, etc.
 
-We propose keeping the default behavior of `Assembly.Location`. That is, 
-
-* If the assembly is loaded directly from the bundle, return empty-string
-* If the assembly is extracted to disk, return the location of the extracted file.
+Proposed solution is for `Assembly.Location` to return the empty-string for bundled assemblies, which is the default behavior for assemblies loaded from memory.
 
 Most of the app development can be agnostic to whether the app is published as single-file or not. However, the parts of the app that deal with physical locations of files need to be aware of the single-file packaging. 
 
 #### `AppContext.BaseDirectory`
 
-A few options to consider for the `AppContext.BaseDirectory` are:
+`AppContext.BaseDirectory` will be the directory where the AppHost (the single-file bundle itself) resides.
 
-* The directory where embedded files are extracted out: This option enables easy access to content files spilled to disk. However, if no files need to be extracted to disk for an application bundle, there's no extraction directory. 
-* The extraction-directory if files are extracted, the AppHost directory otherwise. The limitation to this approach is that the value of  `AppContext.BaseDirectory`  changes with the way applications are packaged/executed.
-* The directory where `AppHost` binary resides (always).
+This is a change from [.net core 3 semantics](design_3_0.md#API-Impact) for single-file apps, where `AppContext.BaseDirectory` returned the directory where the bundle-contents were extracted.  In .net 5, extraction is not performed by default. Therefore, the details about extraction directory are not exposed through the `AppContext.BaseDirectory` API.  
 
-We propose that `AppContext.BaseDirectory`  should always be set to the directory where the `AppHost` bundle resides.  This scheme doesn't provide an obvious mechanism to access the contents of the extraction directory -- by design. The recommended method for accessing content files from the bundle are: 
+However, if files not handled by the runtime are bundled into the single-file (due to `IncludeAllContentInSingleFile` property which emulates .net core 3 bundling behavior)  the app will need a (new) API to locate the extraction directory.
```

Comment:
Since `IncludeAllContentInSingleFile` is meant to replicate .NET Core 3 behavior, another option is to keep the .NET Core 3 behavior where BaseDirectory was modified to point to the extract location.

Note that even if you introduce a new API, it may not be easy for the apps to switch to it: The API may be called in the libraries that the app does not have a control over, etc.

Reply:
I thought about this; but it may be very confusing if the `AppContext.BaseDirectory` changes semantics based on this build property. 

The meaning of `AppContext.BaseDirectory`  is already different based on the TFM. 
An app will likely be built as a single-file using `IncludeAllContentInSingleFile` or not, but not both ways. So, it can handle the correct meaning of `AppContext.BaseDirectory` in its own code. 

However, if a dependent DLL checks for `AppContext.BaseDirectory`, it may interpret the meaning according to `AppContext.TargetFrameworkName`.  But if the directory is different based on the build flags, it will be more confusing.

Reply:
`AppContext.BaseDirectory` differs based on the host. Different hosts tend to configure it differently to make stuff work.

From the `AppContext.BaseDirectory` docs: "Gets the base directory that the assembly resolver uses to probe for assemblies.". It is an oversimplification because the assembly resolver uses multiple directories frequently. So you pick the one that is most likely going to stuff work.

> `IncludeAllContentInSingleFile`

If we going to have this option, I think we should look at it as .NET Core 3 single file compatibility option, with all that comes with it.

Reply:
Thanks for the clarification @jkotas. I have updated the proposal now.